### PR TITLE
limit_req_zone with 4 arguments not being analyzed

### PR DESCRIPTION
**Pull request by oliveromahony**
### Proposed changes

Limit request zone has an issue with not being throwing: invalid number of arguments in \"limit_req_zone\" directive error despite being 4 arguments. 

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [`CONTRIBUTING`](https://github.com/nginxinc/nginx-go-crossplane/blob/main/CONTRIBUTING.md) document
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works
- [x] If applicable, I have checked that any relevant tests pass after adding my changes
- [x] I have updated any relevant documentation ([`README.md`](https://github.com/nginxinc/nginx-go-crossplane/blob/main/README.md))

### CODE REVIEW

File: analyze_test.go [hunk 1]
```diff
@@ -603,6 +603,43 @@ func TestAnalyze_nap_app_protect_enable(t *testing.T) {
...
+			false,
+		},
+	}
+
+	for name, tc := range testcases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			err := analyze("nginx.conf", tc.stmt, ";", tc.ctx, &ParseOptions{
+				DirectiveSources: []MatchFunc{NgxPlusLatestDirectivesMatchFn, AppProtectWAFv4DirectivesMatchFn},
```
Comment (ryepup) [inline analyze_test.go:629]: I think this test would pass without any other changes if we use a different source

```suggestion
				DirectiveSources: []MatchFunc{NgxPlusR31DirectivesMatchFn, AppProtectWAFv4DirectivesMatchFn},
```

`NgxPlusLatestDirectivesMatchFn` thinks `limit_req_zone` has a mask of `ngxHTTPMainConf | ngxConfTake3 | ngxConfTake2,`, but `NgxPlusR31DirectivesMatchFn` has the correct mask `ngxHTTPMainConf | ngxConfTake3 | ngxConfTake4`

Comment (ornj) [reply analyze_test.go:629]: Adding details from our conversation external to the PR. `Take4` is only correct for NGINX+ to allow the `sync` flag. OSS is `Take3`.

**Comment by dengsh12** [comment]
The problem comes from a mismatch in a legacy version of code generator, the flag `NGX_CONF_TAKE4` of `keyval` and `limit_req_zone` mapped to `ngxConfTake2` wrongly. `analyze_nplus_latest_directives.go` may be override to the legacy version while I was dealing with branches. I've proposed another MR #105 to fix. Thanks for your findings!

For the `uint` part it seems we define the bitmasks using `untyped int` and they can be compatible with `uint`, maybe we don't need to change them.

**Comment by oliveromahony** [comment]
> The problem comes from a mismatch in a legacy version of code generator, the flag `NGX_CONF_TAKE4` of `keyval` and `limit_req_zone` mapped to `ngxConfTake2` wrongly. `analyze_nplus_latest_directives.go` may be override to the legacy version while I was dealing with branches. I've proposed another MR #105 to fix. Thanks for your findings!
> 
> For the `uint` part it seems we define the bitmasks using `untyped int` and they can be compatible with `uint`, maybe we don't need to change them.

Happy to drop close this PR. Can we add tests to cover this scenario in  https://github.com/nginxinc/nginx-go-crossplane/pull/105 ?

**Comment by dengsh12** [comment]
> > The problem comes from a mismatch in a legacy version of code generator, the flag `NGX_CONF_TAKE4` of `keyval` and `limit_req_zone` mapped to `ngxConfTake2` wrongly. `analyze_nplus_latest_directives.go` may be override to the legacy version while I was dealing with branches. I've proposed another MR #105 to fix. Thanks for your findings!
> > For the `uint` part it seems we define the bitmasks using `untyped int` and they can be compatible with `uint`, maybe we don't need to change them.
> 
> Happy to close this PR. Can we add tests to cover this scenario in #105 ?

Yeah! Root reason comes from a legacy version of the generator has a wrong mapping from `NGX_CONF_TAKE4` to `ngxConfTake2` and I added [one unit test](https://github.com/nginxinc/nginx-go-crossplane/blob/e00ed7072a655fa52237f7c1500e6f0aaec51318/internal/generator/generator_util_test.go#L112C19-L112C43) with all bitmask in nginx to the MR of generator  #101 

It would also be helpful to add an unit test for `limit_req_zone`. Your test case in this PR offers a very good example, could I add it to #105 ?

**Comment by ryepup** [comment]
> It would also be helpful to add an unit test for `limit_req_zone`. Your test case in this PR is a very good example, could I add it to #105 ?

please do

**Comment by oliveromahony** [comment]
> > > The problem comes from a mismatch in a legacy version of code generator, the flag `NGX_CONF_TAKE4` of `keyval` and `limit_req_zone` mapped to `ngxConfTake2` wrongly. `analyze_nplus_latest_directives.go` may be override to the legacy version while I was dealing with branches. I've proposed another MR #105 to fix. Thanks for your findings!
> > > For the `uint` part it seems we define the bitmasks using `untyped int` and they can be compatible with `uint`, maybe we don't need to change them.
> > 
> > 
> > Happy to close this PR. Can we add tests to cover this scenario in #105 ?
> 
> Yeah! Root reason comes from a legacy version of the generator has a wrong mapping from `NGX_CONF_TAKE4` to `ngxConfTake2` and I added [one unit test](https://github.com/nginxinc/nginx-go-crossplane/blob/e00ed7072a655fa52237f7c1500e6f0aaec51318/internal/generator/generator_util_test.go#L112C19-L112C43) with all bitmask in nginx to the MR of generator #101
> 
> It would also be helpful to add an unit test for `limit_req_zone`. Your test case in this PR is a very good example, could I add it to #105 ?

That would be great

**Pull request closed without merging**